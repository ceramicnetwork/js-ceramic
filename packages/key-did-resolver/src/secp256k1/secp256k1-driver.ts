import { Secp256k1KeyPair } from "./secp256k1-key-pair"

/**
 * Constructs the document based on the method key
 */
const keyToDidDoc = (fingerprint: string): any => {
    const secp256k1KeyPair = Secp256k1KeyPair.fromFingerprint(fingerprint)
    const did = `did:key:${secp256k1KeyPair.fingerprint()}`
    const keyId = `#${secp256k1KeyPair.fingerprint()}`
    return {
        '@context': ['https://www.w3.org/ns/did/v1', {
            '@base': did,
        },],
        id: did,
        publicKey: [{
            id: keyId,
            type: secp256k1KeyPair.type,
            controller: did,
            publicKeyBase58: secp256k1KeyPair.publicKeyBase58,
        },],
        authentication: [keyId],
        assertionMethod: [keyId],
        capabilityDelegation: [keyId],
        capabilityInvocation: [keyId],
    }
}

/**
 * Resolves single DID
 */
const resolve = async ({ did }: any = {}): Promise<object> => {
    if (!did) {
        throw new TypeError('"did" must be a string.');
    }
    const fingerprint = did
        .split('#')[0]
        .split('did:key:')
        .pop();
    return keyToDidDoc(fingerprint);
}

export {
    keyToDidDoc, resolve
}
