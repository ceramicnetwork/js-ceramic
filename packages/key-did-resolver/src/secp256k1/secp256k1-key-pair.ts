import crypto from 'crypto'

import bs58 from 'bs58'
import secp256k1 from 'secp256k1'
import * as utils from './secp256k1-utils'

/**
 * Encapsulates Secp256k1 key pair
 */
export class Secp256k1KeyPair {
    public id: string
    public type: string
    public controller: string

    public publicKeyBase58: string
    public privateKeyBase58: string

    constructor(options: any = {}) {
        this.type = 'EcdsaSecp256k1VerificationKey2019'

        this.id = options.id
        this.controller = options.controller
        this.privateKeyBase58 = options.privateKeyBase58
        this.publicKeyBase58 = options.publicKeyBase58
    }

    /**
     * Constructs fingerprint from public key
     */
    static fingerprintFromPublicKey({ publicKeyBase58 }: any): string {
        const pubkeyBytes = bs58.decode(publicKeyBase58)
        const buffer = new Uint8Array(2 + pubkeyBytes.length)
        // See https://github.com/multiformats/multicodec/blob/master/table.csv
        // 0xe7 is Secp256k1 public key
        buffer[0] = 0xe7 //
        buffer[1] = 0x01
        buffer.set(pubkeyBytes, 2)
        // prefix with `z` to indicate multi-base base58btc encoding
        return `z${bs58.encode(buffer)}`
    }

    /**
     * Derive instance from options
     */
    static async from(options: any): Promise<Secp256k1KeyPair> {
        let privateKeyBase58 = options.privateKeyBase58
        let publicKeyBase58 = options.publicKeyBase58

        if (options.privateKeyHex) {
            privateKeyBase58 = await utils.privateKeyBase58FromPrivateKeyHex(options.privateKeyHex)
        }

        if (options.publicKeyHex) {
            publicKeyBase58 = await utils.publicKeyBase58FromPublicKeyHex(options.publicKeyHex)
        }

        if (options.privateKeyJwk) {
            privateKeyBase58 = utils.privateKeyBase58FromPrivateKeyHex(await utils.privateKeyHexFromJwk(options.privateKeyJwk))
        }

        if (options.publicKeyJwk) {
            publicKeyBase58 = utils.privateKeyBase58FromPrivateKeyHex(await utils.publicKeyHexFromJwk(options.publicKeyJwk))
        }

        return new Secp256k1KeyPair({
            ...options, privateKeyBase58, publicKeyBase58,
        })
    }

    /**
     * Derive instance from fingerprint
     */
    static fromFingerprint(fingerprint: any): Secp256k1KeyPair {
        // skip leading `z` that indicates base58 encoding
        const buffer = bs58.decode(fingerprint.substr(1))

        // https://github.com/multiformats/multicodec/blob/master/table.csv#L77
        if (buffer[0] === 0xe7 && buffer[1] === 0x01) {
            const publicKeyBase58 = bs58.encode(buffer.slice(2))
            const did = `did:key:${Secp256k1KeyPair.fingerprintFromPublicKey({
                publicKeyBase58,
            })}`
            const keyId = `#${Secp256k1KeyPair.fingerprintFromPublicKey({
                publicKeyBase58,
            })}`
            return new Secp256k1KeyPair({
                id: keyId, controller: did, publicKeyBase58,
            })
        }

        throw new Error(`Unsupported Fingerprint Type: ${fingerprint}`)
    }

    /**
     * Get public key
     */
    get publicKey(): string {
        return this.publicKeyBase58
    }

    /**
     * Get signer functionality
     */
    signer(): any {
        if (!this.privateKeyBase58) {
            return {
                async sign(): Promise<void> {
                    throw new Error('No private key to sign with.')
                },
            }
        }
        const privateKeyBase58 = this.privateKeyBase58
        return {
            async sign({ data }: any): Promise<string> {
                const messageHashUInt8Array = crypto
                    .createHash('sha256')
                    .update(data)
                    .digest()
                const privateKeyUInt8Array = await utils.privateKeyUInt8ArrayFromPrivateKeyBase58(privateKeyBase58)
                const sigObj: any = secp256k1.ecdsaSign(messageHashUInt8Array, privateKeyUInt8Array)

                return sigObj.signature
            },
        }
    }

    /**
     * Get fingerprint from the instance
     */
    fingerprint(): string {
        const { publicKeyBase58 } = this
        return Secp256k1KeyPair.fingerprintFromPublicKey({ publicKeyBase58 })
    }

    /**
     * To JWK
     * @param _private
     */
    toJwk(_private = false): string {
        if (_private) {
            return utils.privateKeyJwkFromPrivateKeyHex(bs58.decode(this.privateKeyBase58).toString('hex'))
        }
        return utils.publicKeyJwkFromPublicKeyHex(bs58.decode(this.publicKeyBase58).toString('hex'))
    }

}
