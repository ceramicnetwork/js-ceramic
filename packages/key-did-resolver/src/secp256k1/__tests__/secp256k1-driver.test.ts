import { Secp256k1KeyPair } from "../secp256k1-key-pair"
import * as driver from "../secp256k1-driver"

describe('Secp256k1 driver', () => {

    it('successfully creates Secp256k1KeyPair instance from fingerprint', async () => {
        const keyPair = Secp256k1KeyPair.fromFingerprint("zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV")
        expect(keyPair).toMatchSnapshot()
    })

    it('successfully resolves the document from did', async () => {
        const document = await driver.resolve({ did: "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV" })
        expect(document).toMatchSnapshot()
    })

})

