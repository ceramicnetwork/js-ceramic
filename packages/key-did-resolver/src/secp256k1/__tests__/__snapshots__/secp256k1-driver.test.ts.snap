// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secp256k1 driver successfully creates Secp256k1KeyPair instance from fingerprint 1`] = `
Secp256k1KeyPair {
  "controller": "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  "id": "#zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  "privateKeyBase58": undefined,
  "publicKeyBase58": "2Bv3fQSJiGJsbUUfUdiPsXXZGG5cXqaSeYnieuJcUFAbm",
  "type": "EcdsaSecp256k1VerificationKey2019",
}
`;

exports[`Secp256k1 driver successfully resolves the document from did 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/ns/did/v1",
    Object {
      "@base": "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    },
  ],
  "assertionMethod": Array [
    "#zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  ],
  "authentication": Array [
    "#zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  ],
  "capabilityDelegation": Array [
    "#zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  ],
  "capabilityInvocation": Array [
    "#zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  ],
  "id": "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
  "publicKey": Array [
    Object {
      "controller": "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
      "id": "#zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
      "publicKeyBase58": "2Bv3fQSJiGJsbUUfUdiPsXXZGG5cXqaSeYnieuJcUFAbm",
      "type": "EcdsaSecp256k1VerificationKey2019",
    },
  ],
}
`;
